presto-spi/src/main/java/com/facebook/presto/spi/memory/AggregatedMemoryContext.java

```diff
@@ -11,39 +11,52 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.memory;
+package com.facebook.presto.spi.memory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
 
-import static com.google.common.base.MoreObjects.toStringHelper;
-import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
+@ThreadSafe
 public class AggregatedMemoryContext
-        extends AbstractAggregatedMemoryContext
 {
-    // This class should remain exactly the same as AggregatedMemoryContext in com.facebook.presto.orc.memory and com.facebook.presto.hive.parquet.memory
-
-    private final AbstractAggregatedMemoryContext parentMemoryContext;
+    @Nullable
+    @GuardedBy("this")
+    private final AggregatedMemoryContext parentMemoryContext;
+    @GuardedBy("this")
     private long usedBytes;
+    @GuardedBy("this")
     private boolean closed;
 
     public AggregatedMemoryContext()
     {
         this.parentMemoryContext = null;
     }
 
-    public AggregatedMemoryContext(AbstractAggregatedMemoryContext parentMemoryContext)
+    private AggregatedMemoryContext(AggregatedMemoryContext parentMemoryContext)
     {
         this.parentMemoryContext = requireNonNull(parentMemoryContext, "parentMemoryContext is null");
     }
 
-    public long getBytes()
+    public AggregatedMemoryContext newAggregatedMemoryContext()
+    {
+        return new AggregatedMemoryContext(this);
+    }
+
+    public LocalMemoryContext newLocalMemoryContext()
+    {
+        return new LocalMemoryContext(this);
+    }
+
+    public synchronized long getBytes()
     {
         checkState(!closed);
         return usedBytes;
     }
 
-    @Override
-    protected void updateBytes(long bytes)
+    public synchronized void updateBytes(long bytes)
```

Comment:
The original design of memory context is so that callers do absolute (instead of relative) updates. It is a specific choice that updateBytes in AggregatedMemoryContext have a limited visibility scope.

If you need to implement some kind of force-free, it should not need to access `updateBytes` directly because `close` is available.